src/coreclr/vm/util.hpp

```diff
@@ -918,6 +918,86 @@ class COMCharacter {
     static BOOL nativeIsDigit(WCHAR c);
 };
 
+// ======================================================================================
+// Simple, reusable 100ns timer for normalizing ticks. For use in Q/FCalls to avoid discrepency with
+// tick frequency between native and managed.
+class NormalizedTimer
+{
+private:
+    LARGE_INTEGER startTimestamp = { .QuadPart = 0 };
+    LARGE_INTEGER stopTimestamp = { .QuadPart = 0 };
+    int64_t cachedElapsed100nsTicks = 0;
+    bool shouldRecalculate = true;
+    bool isRunning = false;
+    static const int64_t NormalizedTicksPerSecond = 10000000 /* 100ns ticks per second (1e7) */;
+    static Volatile<int64_t> s_frequency;
+
+    inline
+    void Lap()
+    {
+        startTimestamp = stopTimestamp;
+        stopTimestamp.QuadPart = 0;
+        isRunning = true;
+        shouldRecalculate = true;
+    }
+public:
+    // ======================================================================================
+    // Start the timer
+    inline
+    void Start()
+    {
+        QueryPerformanceCounter(&startTimestamp);
+        stopTimestamp.QuadPart = 0;
+        isRunning = true;
+        shouldRecalculate = true;
+    }
+
+    // ======================================================================================
+    // stop the timer. If called before starting, sets the start time to the same as the stop
+    inline
+    void Stop()
+    {
+        QueryPerformanceCounter(&stopTimestamp);
+        // protect against stop before start
+        if (!isRunning)
+            startTimestamp = stopTimestamp;
+
+        isRunning = false;
+        shouldRecalculate = true;
+    }
+
+    // ======================================================================================
+    // Return elapsed ticks. This will stop a running timer.
+    // Will return 0 if called out of order.
+    // Only recalculated this value if it has been stopped/started since previous calculation.
+    inline
+    int64_t Elapsed100nsTicks(bool shouldContinue = false)
+    {
+        if (s_frequency == -1)
+        {
+            int64_t frequency;
+            LARGE_INTEGER qpfValue;
+            QueryPerformanceFrequency(&qpfValue);
+            frequency = static_cast<int64_t>(qpfValue.QuadPart);
+            frequency /= NormalizedTicksPerSecond;
```

Comment:
Can frequency be less than NormalizedTicksPerSecond here?

Reply:
If the resolution of the clock used by `QueryPerformanceFrequency` is larger than 100ns, then yes, `frequency` could be smaller than `NormalizedTicksPerSecond`.

The [docs](https://docs.microsoft.com/en-us/windows/win32/sysinfo/acquiring-high-resolution-time-stamps#resolution-precision-accuracy-and-stability) specifically have a sample where the resolution is 320ns on a machine. I think I should change frequency to a floating-point value to we handle those situations.

On non-Windows platforms, the CoreCLR PAL version of `WueryPerformanceFrequency` seems to always return 1e9 (ns in a s).

https://github.com/dotnet/runtime/blob/fe9a54df8f4b5e6a2e4cafc0e4234a64e8c9cf8e/src/coreclr/pal/src/misc/time.cpp#L196-L223